Datarow tests - support methods with optional parameters

ssa3512 opened the pull request:
Resolves #252

msftclas commented:
[![CLA assistant check](https://cla.opensource.microsoft.com/pull/badge/not_signed)](https://cla.opensource.microsoft.com/Microsoft/testfx?pullRequest=604) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.opensource.microsoft.com/Microsoft/testfx?pullRequest=604) before we can accept your contribution.<br/><br/>:x: ssa3512 [sign now](https://cla.opensource.microsoft.com/Microsoft/testfx?pullRequest=604)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.opensource.microsoft.com/check/Microsoft/testfx?pullRequest=604) it.</sub>

jayaranigarg commented on src/Adapter/MSTest.CoreAdapter/Execution/TestMethodInfo.cs:
```diff
@@ -171,7 +171,48 @@ public virtual TestResult Invoke(object[] arguments)
+                return arguments;
+            }
+
+            if (arguments == null)
```
Should we just add this check to SetArguments() function itself and call ResolveArguments() only when `arguments` is not null?

ssa3512 replied:
moved null check to SetArguments()

jayaranigarg commented on src/Adapter/MSTest.CoreAdapter/Execution/TestMethodInfo.cs:
```diff
@@ -171,7 +171,48 @@ public virtual TestResult Invoke(object[] arguments)
+                return null;
+            }
+
+            if (arguments.Length < requiredParameterCount || arguments.Length > parameterInfos.Length)
```
Please add explanatory comments wherever possible.

ssa3512 replied:
added comments in https://github.com/Microsoft/testfx/pull/604/commits/65cd652b25348974bc42a8e3ae57416af708f23c

jayaranigarg commented on src/Adapter/MSTest.CoreAdapter/Execution/TestMethodInfo.cs:
```diff
@@ -171,7 +171,48 @@ public virtual TestResult Invoke(object[] arguments)
+            this.arguments = this.ResolveArguments(arguments);
+        }
+
+        internal object[] ResolveArguments(object[] arguments)
```
Please add UnitTests for this function.

ssa3512 replied:
added tests in TestMethodInfoTests.cs

jayaranigarg commented on test/E2ETests/Smoke.E2E.Tests/DataRowTests.cs:
```diff
@@ -21,12 +21,31 @@ public void ExecuteOnlyDerivedClassDataRowsWhenBothBaseAndDerviedClassHasDataRow
                 "DataRowTestMethod (BaseString3)",
                 "DataRowTestMethod (DerivedString1)",
-                "DataRowTestMethod (DerivedString2)");
+                "DataRowTestMethod (DerivedString2)",
```
This check has become quite long. It would be more readable if we have seperate E2E tests for `DataRowTestMethodWithSomeOptionalParameters` and `DataRowTestMethodWithAllOptionalParameters`

jayaranigarg commented on test/E2ETests/TestAssets/DataRowTestProject/BaseClass.cs:
```diff
@@ -16,5 +16,26 @@ public virtual void DataRowTestMethod(string a)
+        [DataRow(42, "BaseOptionalString1")]
+        [DataRow(42, "BaseOptionalString2", "BaseOptionalString3")]
+        [DataRow(42, "BaseOptionalString4", "BaseOptionalString5")]
+        public virtual void DataRowTestMethodWithSomeOptionalParameters(int i, string s1 = null, string s2 = null)
```
Please add negative test cases as well, something where user specifies less than required arguments or more than required arguments and as a result of which the test method fails with appropriate exception.

ssa3512 replied:
added with DataRowOptionalInvalidArguments

jayaranigarg commented:
@ssa3512 : Thank you for your contribution!! 
PR looks good overall. Kindly do the needful changes so that we can push this in.

jayaranigarg commented:
@ssa3512 : It would be even more wonderful if you can address this related issue https://github.com/Microsoft/testfx/issues/283 as well as part of this change.

jayaranigarg commented:
@ssa3512 : Any updates on this? Let us know if you need some help on this.

ssa3512 commented:
@jayaranigarg sorry, no updates yet. I'm hoping to have some time to address the feedback and look at including #283 early next week.

ssa3512 commented:
@jayaranigarg I believe I addressed all feedback; should now resolve #283 as well. Please review again and let me know if anything else needs to be addressed.

ssa3512 commented:
had to rebase - some commits were made with incorrect user id (work email)

ssa3512 commented on src/Adapter/MSTest.CoreAdapter/Execution/TestMethodInfo.cs:
```diff
@@ -171,7 +171,94 @@ public virtual TestResult Invoke(object[] arguments)
+            }
+
+            object[] newParameters = new object[parameterInfos.Length];
+            for (int i = 0; i < arguments.Length; i++)
```
resolved with https://github.com/Microsoft/testfx/pull/604/commits/30b76e08df4679a8fcda4ba7bf90ba247aacf5f9

ssa3512 commented on test/E2ETests/Smoke.E2E.Tests/DataRowTests.cs:
```diff
@@ -23,22 +23,156 @@ public void ExecuteOnlyDerivedClassDataRowsWhenBothBaseAndDerviedClassHasDataRow
+        }
+
+        [TestMethod]
+        public void ExecuteOnlyDerivedClassDataRowsWhenBothBaseAndDerviedClassHasDataRows_DataRowSomeOptional()
```
Resolved with https://github.com/Microsoft/testfx/pull/604/commits/30b76e08df4679a8fcda4ba7bf90ba247aacf5f9

jayaranigarg commented on test/E2ETests/Smoke.E2E.Tests/DataRowTests.cs:
```diff
@@ -23,23 +23,102 @@ public void ExecuteOnlyDerivedClassDataRowsWhenBothBaseAndDerviedClassHasDataRow
+                "DataRowTestMethodWithSomeOptionalParameters (123,DerivedOptionalString2,DerivedOptionalString3)");
+
+            // 5 tests of BaseClass.DataRowTestMethodWithSomeOptionalParameters - 4 datarow result and 1 parent result
+            // 4 tests of DerivedClass.DataRowTestMethodWithSomeOptionalParameters - 3 datarow result and 1 parent result
```
@ssa3512 : We can simply these tests even more by just keeping the `DataRowTestMethodWithSomeOptionalParameters` in the base class only.

Apologies, I should have been more clear.

jayaranigarg commented on test/E2ETests/Smoke.E2E.Tests/DataRowTests.cs:
```diff
@@ -23,23 +23,102 @@ public void ExecuteOnlyDerivedClassDataRowsWhenBothBaseAndDerviedClassHasDataRow
+                "DataRowTestMethodWithAllOptionalParameters (42,BaseOptionalString6)",
+                "DataRowTestMethodWithAllOptionalParameters (42,BaseOptionalString7,BaseOptionalString8)",
+                "DataRowTestMethodWithAllOptionalParameters (42,BaseOptionalString9,BaseOptionalString10)",
+                "DataRowTestMethodWithAllOptionalParameters (123)",
```
Similar here.

jayaranigarg commented on test/E2ETests/Smoke.E2E.Tests/DataRowTests.cs:
```diff
@@ -23,23 +23,102 @@ public void ExecuteOnlyDerivedClassDataRowsWhenBothBaseAndDerviedClassHasDataRow
+                "DataRowTestMethodWithParamsParameters (1,BaseSingleParamsArg)",
+                "DataRowTestMethodWithParamsParameters (1,BaseParamsArg1,BaseParamsArg2)",
+                "DataRowTestMethodWithParamsParameters (1,BaseParamsArg1,BaseParamsArg2,BaseParamsArg3)",
+                "DataRowTestMethodWithParamsParameters (2)",
```
Same

jayaranigarg commented on test/E2ETests/TestAssets/DataRowTestProject/DerivedClass.cs:
```diff
@@ -8,12 +8,53 @@ namespace DataRowTestProject
             Assert.IsTrue(true);
         }
+
+        [TestCategory("DataRowSomeOptional")]
```
We can remove these testmethods from derived class.

jayaranigarg commented on test/E2ETests/Smoke.E2E.Tests/DataRowTests.cs:
```diff
@@ -23,23 +23,102 @@ public void ExecuteOnlyDerivedClassDataRowsWhenBothBaseAndDerviedClassHasDataRow
+                "DataRowTestMethodFailsWithInvalidArguments ()",
+                "DataRowTestMethodFailsWithInvalidArguments (1)",
+                "DataRowTestMethodFailsWithInvalidArguments (1,BaseRequiredArgument,BaseOptionalArgument,BaseExtraArgument)",
+                "DataRowTestMethodFailsWithInvalidArguments (2)",
```
Here as well.

ssa3512 commented:
@jayaranigarg if we remove the overridden tests from DerivedClass, the E2E tests run the base class tests twice - once for BaseClass and once for DerivedClass, causing the counts of passed tests to be doubled. Is it okay to specify `FullyQualifiedName~BaseClass` in the testCaseFilter for the new E2E tests to prevent this, causing only the base class to run, or should I break the new tests out into a separate DataRowTest class that does not have a derived class?

jayaranigarg commented:
> @jayaranigarg if we remove the overridden tests from DerivedClass, the E2E tests run the base class tests twice - once for BaseClass and once for DerivedClass, causing the counts of passed tests to be doubled. Is it okay to specify `FullyQualifiedName~BaseClass` in the testCaseFilter for the new E2E tests to prevent this, causing only the base class to run, or should I break the new tests out into a separate DataRowTest class that does not have a derived class?

@ssa3512 : Oh I see. I believe if we just keep these tests in the derived class, then they should not be executed twice.

ssa3512 commented:
> @ssa3512 : Oh I see. I believe if we just keep these tests in the derived class, then they should not be executed twice.

@jayaranigarg In that case is everything good to go?

jayaranigarg commented on test/E2ETests/TestAssets/DataRowTestProject/BaseClass.cs:
```diff
@@ -16,5 +17,50 @@ public virtual void DataRowTestMethod(string a)
+        [DataRow(42, "BaseOptionalString6")]
+        [DataRow(42, "BaseOptionalString7", "BaseOptionalString8")]
+        [DataRow(42, "BaseOptionalString9", "BaseOptionalString10")]
+        public virtual void DataRowTestMethodWithAllOptionalParameters(int i = 0, string s1 = null, string s2 = null)
```
We can remove these test from the **base** class

jayaranigarg commented:
@ssa3512 : I have updated my comments. You can remove the tests from the base class and have them in derived class only. It will be good to go after that.

ssa3512 commented:
@jayaranigarg removed tests from base class and added the `[DataRow()]` tests that were only in base to Derived

jayaranigarg commented:
@ssa3512 : Looks nice🎉 I will go ahead and push this in. Thank you for the contributions 🥇

jnyrup commented:
@jayaranigarg Should #283 be closed as well?

jayaranigarg commented:
@jnyrup: Yes. This PR adds support for params as well. Updated the title to be more explicit.

julealgon commented:
`params` support is completely unrelated to supporting default parameter values.

I'd suggest updating the release notes to make it clear that `params` is also supported on 2.0.0 now. The current release notes doesn't mention it at all, I assume because it only keeps track of issues, and we didn't have an issue to track this part of the work here:

![image](https://user-images.githubusercontent.com/461579/68244783-54b8a180-fff4-11e9-9499-ff7807449dea.png)

In the future, I'd suggest being mindful of these changes and:
1. Sending them in separate PRs to promote isolation of changes
2. Ensuring that all functional changes are covered by an issue item for tracking
